accepted/2024/[TS]PMUI-Readme-rendering.md

```diff
@@ -0,0 +1,102 @@
+
+We can use the Preview builder as follows:
+```C#
+//This creates a new instance of the preview builder
+var markdownPreview = new PreviewBuilder().Build();
+
+//We update the current markdown being rendered by calling "UpdateContentAsync"
+markdownPreview.UpdateContentAsync(markDown ?? string.Empty, ScrollHint.None)
+
+//IMarkdownPreview.VisualElement contains the FrameworkElement to be passed to the view
+MarkdownPreviewControl = markdownPreview.VisualElement
+```
+#### Locating the ReadMe
+For packages that are already on the disk we will retrieve the ReadMe from the root folder of the package. 
+
+For remote packages we will download the file and pull the md from the stream response using the [ZipArchive](https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.ziparchive?view=net-8.0) class.
+
+If the feed we're calling allows for [HTTP range requests](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests) we can implement something similar to [Seekable HTTP Range Stream](https://codereview.stackexchange.com/questions/70679/seekable-http-range-stream) to reduce the amount of data we pull. 
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+
+## Drawbacks
+
+This control is not currently using Daytona. There are plans to upgrade it, [Experience 1870733](https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1870733), but they have been delayed. 
+
+It's also currently marked as obsolete since the interface has not been finalized. So when an upgrade is made we may have to change how we use the control.
+
+Currently WebView2 controls always render ontop of other controls in the view. [Secnario 25254665](https://microsoft.visualstudio.com/Edge/_workitems/edit/25254665). PM UI needs to be updated to ensure items don't scroll off screen.
+
+## Rationale and alternatives
+By using an existing control we maintain consistency throughout the IDE and can rely on the owner to fix any bugs with the control.
+<!-- Why is this the best design compared to other designs? -->
+<!-- What other designs have been considered and why weren't they chosen? -->
+<!-- What is the impact of not doing this? -->
+
+## Prior Art
+The IMarkdownPreview is currently being used when creating a new pull request inside of Visual Studio.  
+![Alt text](../../meta/resources/ReadMePMUI/PullRequestExperience.png) 
+<!-- What prior art, both good and bad are related to this proposal? -->
+<!-- Do other features exist in other ecosystems and what experience have their community had? -->
+<!-- What lessons from other communities can we learn from? -->
+<!-- Are there any resources that are relevant to this proposal? -->
+
+## Unresolved Questions
```

Comment:
Will non-embedded readmes be supported? As mentioned on the teams threads, a package author can provide a readme outside of the .nupkg and this is visible on NuGet.org. It's also mutable. I've seen indication from customers in the past that this is a desirable feature as an alternative to embedded readmes since it allows fixups later and, for these users, readme authoring is a step sometime after pack is done in their desired workflow. I think @clairernovotny may have thoughts+data on this. Despite my feeling, a large number of readmes are non-embedded:

Counts on nuget.org.
```kusto
NiPackageReadmes
| where ResultType !in ("Deleted", "None")
| summarize count() by ResultType
```

|ResultType|count_|
|---|---|
|Embedded|748714|
|Legacy|248580|

Reply:
Is there only one version of the non-embedded files ReadMe for the package? 

As it stands this plan was only for Embedded ReadMe, seeing as we're going to focus on packages that are already downloaded. We would prefer to have a URL provided by server allowing us to download the MD for packages which aren't downloaded. We don't want to download the package temporarily due to perf concerns. That URL could return the non-embedded ReadMe and Client wouldn't know the difference.

Reply:
Non-embedded readme and embedded readme are both per-version. 

Server could theoretically supply the non-embedded README in the URL, and yes client wouldn't know the difference. The only problem is that off of the Browse tab, like on the installed tab there is not necessarily server metadata available to get the non-embedded readme. But this could maybe be resolved by fetching it from the package source in a new metadata fetching flow. Not sure. There are options.

Reply:
Take an analogy of embedded license and embedded icon. The legacy counterparts have the content available externally but it still sort of works. The data is available somehow.

For old icon URL (not embedded icons), nuget.org caches the icon and surfaces them in storage and API endpoints. They look just like embedded icons if you look and JSON returned from our APIs.

For licenses, you can still click through to the license URL for such cases. In other words, graceful fallback.

It feels a little strange to not have the legacy readmes available. Maybe a link to the readme on NuGet.org (not downloading the markdown) as the graceful fallback is enough.